object.go

```diff
@@ -578,77 +613,82 @@ func (obs *obs) Delete(name string) error {
 		return ErrBadObjectMeta
 	}
 
-	// Place a rollup delete marker.
+	// Place a rollup delete marker / publish the meta
 	info.Deleted = true
 	info.Size, info.Chunks, info.Digest = 0, 0, _EMPTY_
-
-	metaSubj := fmt.Sprintf(objMetaPreTmpl, obs.name, sanitizeName(name))
-	mm := NewMsg(metaSubj)
-	mm.Data, err = json.Marshal(info)
-	if err != nil {
+	if err = publishMeta(obs, name, info); err != nil {
 		return err
 	}
-	mm.Header.Set(MsgRollup, MsgRollupSubject)
-	_, err = obs.js.PublishMsg(mm)
+
+	// Purge chunks for the object
+	return purgeChunks(obs, info.NUID)
+}
+
+func doesntExistOrIsDeleted(obs *obs, name string) error {
+	info, err := obs.GetInfo(name)
 	if err != nil {
+		if err == ErrObjectNotFound {
+			return nil
+		}
 		return err
 	}
+	if info.Deleted {
+		return nil
+	}
 
-	// Purge chunks for the object.
-	chunkSubj := fmt.Sprintf(objChunksPreTmpl, obs.name, info.NUID)
-	return obs.js.purgeStream(obs.stream, &streamPurgeRequest{Subject: chunkSubj})
+	return errors.New("nats: name already used for an existing object")
 }
 
-// AddLink will add a link to another object into this object store.
+// AddLink will add a link to another object.
 func (obs *obs) AddLink(name string, obj *ObjectInfo) (*ObjectInfo, error) {
-	if obj == nil {
+	if obj == nil || obj.Name == "" || obj.Bucket == "" {
 		return nil, errors.New("nats: object required")
 	}
-	if obj.Deleted {
-		return nil, errors.New("nats: object is deleted")
-	}
-	name = sanitizeName(name)
-	if !keyValid(name) {
-		return nil, ErrInvalidObjectName
-	}
 
-	// Same object store.
-	if obj.Bucket == obs.name {
-		info := *obj
-		info.Name = name
-		if err := obs.UpdateMeta(obj.Name, &info.ObjectMeta); err != nil {
-			return nil, err
-		}
-		return obs.GetInfo(name)
-	}
+	// If the link name already exists in this store, you cannot add as a link
+	// If the link name exists as a deleted object, that's fine
+	// @derek Maybe we need an addOrUpdateLink or just an updateLink function,
+	// or add a flag/option to the method?
+	// either way, we should not update a non link to a link
```

Comment:
What happens if you put something in under same name today? Meaning behavior of non-link put when something already there?

Reply:
the previous add link would do different things if the link was bucket to same bucket or bucket to different bucket.

bucket to same bucket: this code had errors. 

so I wasn't sure what original line 617 did. I thought that info and obj now point to the exact same object versus info being a deep copy.
In original line 618 you set that info.Name to the (sanitized) name. I assumed that this also set obj.Name

if info and obj is the same line 619 calls UpdateMeta with some sanitized object name that isn't lookup-able.
if info and obj diverge 619 would overwrite the object you are trying top link to.

bucket to diff bucket. This code also used the sanitized name when it should not have, but otherwise is the foundation for the re-write I did.

Reply:
I actually wanted to change this api to 
```
func (obs *obs) AddLink(name string, targetBucket string, targetName string) (*ObjectInfo, error)
```

Or at least verify the obj that is being sent in, but that would require a getInfo possibly to another bucket understanding that the target bucket may be different.

But that was a lot of work so I landed where I did.

Reply:
I meant non-link normal put when object already exists.

Reply:
So I just looked and realized the put is a problem with the same name. It always makes a new [nu]id. It will clean up (purge) the chunks from the original name's nuid, so if put was called from addlink, it does what you would think and completely replaces the object with a link.
I think the behavior or add link should be like create, also possibly allow overwrite of a name that is a link, but not allow overwrite of a name this is an object.

Reply:
Much like KV, I wonder if we need put/add link with options where the user tells us create/update/allow overwrite, etc. and then what the default behavior is

Reply:
are you thinking that if they want to put something over something else, that's their prerogative?

Reply:
So I'm just simplifying and not doing any checking, just making sure that put does the right think and the add link code is correct. We can always add more granularity later